File: _templates/mustache/render_mustache.py
from __future__ import (absolute_import, division, print_function,
	unicode_literals)

import os, sys, glob, argparse, re, time
from future.builtins import (ascii, str, filter, hex, map, oct, zip)

#import pystache
import chevron

SCRIPTPARENT = os.path.dirname(os.path.abspath(__file__))
CUR_DIR = os.path.abspath(os.curdir)

def deserialize_file(datapath, fmt='yaml', date_key='date'):
	initdata = {}
	
	if fmt in ['yaml', 'json']:
		try:
			import yaml						# [Base|Safe|Full]Loader
			with open(datapath) as fIn:
				initdata.update(yaml.load(fIn, Loader=yaml.SafeLoader))
		except ImportError as exc:
			print(repr(exc))
	elif 'toml' == fmt:
		try:
			import toml
			with open(datapath) as fIn:
				initdata.update(toml.load(fIn))
		except ImportError as exc:
			print(repr(exc))
	#elif 'json' == fmt:
	#	import json
	#	with open(datapath) as fIn:
	#		#initdata = json.load(fIn)
	#		initdata.update(json.load(fIn))
	
	initdata.update({date_key: time.strftime('%Y-%m-%d')})
	return initdata

def deserialize_str(datastr, fmt='yaml', date_key='date'):
	initdata = {}
	
	if datastr is not None:
		if fmt in ['yaml', 'json']:
			try:
				import yaml						# [Base|Safe|Full]Loader
				initdata.update(yaml.load(datastr, Loader=yaml.SafeLoader))
			except ImportError as exc:
				print(repr(exc))
		elif 'toml' == fmt:
			try:
				import toml
				initdata.update(toml.loads(datastr))
			except ImportError as exc:
				print(repr(exc))
		#elif 'json' == fmt:
		#	import json
		#	#initdata = json.loads(datastr.decode(encoding='utf-8'))
		#	initdata.update(json.loads(datastr))
	
	initdata.update({date_key: time.strftime('%Y-%m-%d')})
	return initdata

def regex_checks(pat, substr, txt):
	if not re.match(pat, substr):
		print("ERROR: Regex match failure (re.match({0}, {1})) for ({2}).".format(
			pat, substr, txt))
		sys.exit(1)

def derive_skel_vars(ctx=None):
	name = '{0}{1}{2}'.format(ctx.get('parent', ''), ctx.get('separator', ''),
		ctx['project'])
	parentcap = str.join(ctx.get('joiner', ''), map(lambda e: e.capitalize(),
		ctx.get('parent', '').split(ctx.get('separator', '-'))))
	namespace = '{0}{1}.{2}'.format(ctx['groupid'] + '.' if ctx.get('groupid')
		else '', ctx.get('parent', ''), ctx['project'])
	
	ctx.update({'year': ctx['date'].split('-')[0], 'name': name, 'parentcap': parentcap,
		'projectcap': ctx['project'].capitalize(), 'namespace': namespace,
		'nesteddirs': namespace.replace('.', '/')})
	return ctx

def render_skeleton(skeleton='skeleton-ml', ctx=None):
	ctx.update({'skeletondir': os.path.join(SCRIPTPARENT, skeleton)})
	start_dir = os.path.join(ctx['skeletondir'], '{{name}}')
	files_skel = map(lambda p: os.path.relpath(p, start=start_dir), 
		filter(lambda p: os.path.isfile(p), glob.glob(start_dir + '/**/*',
		recursive=True)  + glob.glob(start_dir + '/**/.*', recursive=True)))
	renderouts, copyouts, pat_mustache = {}, {}, re.compile('\.mustache$')
	for skelX in files_skel:
		if pat_mustache.search(skelX) :
			#renderouts[skelX] = re.sub(pat_mustache, '', pystache.render(
			#	os.path.join(CUR_DIR, ctx['name'], skelX), ctx))
			renderouts[skelX] = re.sub(pat_mustache, '', chevron.render(
				os.path.join(CUR_DIR, ctx['name'], skelX), ctx))
		else:
			#copyouts[skelX] = pystache.render(os.path.join(CUR_DIR, ctx['name'], skelX),
			#	ctx)
			copyouts[skelX] = chevron.render(os.path.join(CUR_DIR, ctx['name'], skelX),
				ctx)
	print('... processing files -- rendering {0} ; copying {1} ...'.format(
		len(renderouts), len(copyouts)))
	
	for dirX in [os.path.dirname(pathX) for pathX in 
			set(renderouts.values()) | set(copyouts.values())]:
		if not os.path.exists(os.path.join(CUR_DIR, ctx['name'], dirX)):
			os.makedirs(os.path.join(CUR_DIR, ctx['name'], dirX))
	for srcR, dstR in renderouts.items():
		with open(os.path.join(CUR_DIR, ctx['name'], dstR), 'w+') as fOut, \
				open(os.path.join(start_dir, srcR)) as fIn:
			#fOut.write(pystache.render(fIn.read(), ctx))
			fOut.write(chevron.render(fIn.read(), ctx))
	for srcC, dstC in copyouts.items():
		with open(os.path.join(CUR_DIR, ctx['name'], dstC), 'w+') as fOut, \
				open(os.path.join(start_dir, srcC)) as fIn:
			fOut.write(fIn.read())
	
	print('Post rendering message')
	os.chdir(ctx['name'])
	os.system('python choices/post_gen_project.py') # python ___.py | sh ___.sh

def parse_cmdopts(args=None):
	opts_parser = argparse.ArgumentParser()
	
	opts_parser.add_argument('template', nargs='?', default='skeleton-ml',
		help='Template path')
	opts_parser.add_argument('-d', '--data', default='data.yaml',
		help='Data path or - (for stdin)')
	opts_parser.add_argument('-f', '--datafmt', default='yaml',
		choices=[None, 'yaml', 'json', 'toml'], help='Specify data file format')
	#opts_parser.add_argument('-i', '--fileIn', type=argparse.FileType('r'),
	#	default=sys.stdin, help='File in - (for stdin) or path')
	opts_parser.add_argument('-o', '--fileOut', type=argparse.FileType('w'),
		default=sys.stdout, help='File out - (for stdout) or path')
	opts_parser.add_argument('-k', '--kvset', metavar='KEY=VALUE', nargs='+',
		help='Set key=value pairs (Ex: -k key1="val1")')
	
	return opts_parser.parse_args(args)

def main(argv=None):
	opts_hash, cfg = parse_cmdopts(argv), {}
	
	if not os.path.exists(opts_hash.template) and \
			not os.path.exists(os.path.join(SCRIPTPARENT, opts_hash.template)):
		print('Non-existent template: {0}'.format(opts_hash.template))
		sys.exit(1)
	kvset = dict(kv.split('=', 1) for kv in opts_hash.kvset) if opts_hash.kvset else None
	is_dir = os.path.isdir(opts_hash.template) or \
			os.path.isdir(os.path.join(SCRIPTPARENT, opts_hash.template))
	if '-' == opts_hash.data:
		cfg = deserialize_str(sys.stdin.read(), fmt=opts_hash.datafmt, date_key='date')
	
	if not is_dir:
		if not '-' == opts_hash.data:
			cfg = deserialize_file(opts_hash.data, fmt=opts_hash.datafmt, date_key='date')
		cfg.update(kvset if kvset else {})
	else:
		if not '-' == opts_hash.data:
			cfg = deserialize_file(os.path.join(SCRIPTPARENT, opts_hash.template,
				opts_hash.data), fmt=opts_hash.datafmt, date_key='date')
		cfg.update(kvset if kvset else {})
		cfg = derive_skel_vars(cfg)
		regex_checks(cfg.get('parentregex', ''), cfg.get('parent', ''), cfg.get('name', ''))
		regex_checks(cfg.get('projectregex', ''), cfg.get('project', ''), cfg.get('name', ''))
		
	switcher = {
		None: lambda: render_skeleton(opts_hash.template, cfg),
		True: lambda: render_skeleton(opts_hash.template, cfg),
		#False: lambda: opts_hash.fileOut.write(pystache.render(
		#	open(opts_hash.template, 'r'), cfg))
		False: lambda: opts_hash.fileOut.write(chevron.render(
			open(opts_hash.template, 'r'), cfg))
	}
	func = switcher.get(is_dir, lambda:
		print('(is_dir: {0}) Invalid template: {1}'.format(is_dir, opts_hash.template)))
	return func()


if '__main__' == __name__:
	import sys
	#raise SystemExit(main(sys.argv[1:]))
	sys.exit(main(sys.argv[1:]))
	

File: _templates/mustache/render_mustache.rb
#!/usr/bin/env ruby -w

require 'English'
require 'json'
require 'psych'
require 'toml'
require 'mustache'
require 'fileutils'

module RenderMustache
	extend self
	
	SCRIPTPARENT = File.dirname(File.absolute_path(__FILE__))

	def deserialize_file(datapath, fmt='yaml', date_key='date')
		initdata = {}
	
		keys2sym = proc {|h| h.map{|k, v|
		  [k.to_sym, v.is_a?(Hash) ? keys2sym.call(v) : v]}.to_h}
	
		if ['yaml', 'json'].include? fmt
			initdata = initdata.merge(keys2sym.call Psych.load_file(datapath))
		elsif 'toml' == fmt
			initdata = initdata.merge(keys2sym.call TOML.load_file(datapath))
		#elsif 'json' == fmt
		#	#initdata = JSON.parse(File.read(datapath), {:symbolize_names => true})
		#	initdata = initdata.merge(JSON.parse(File.read(datapath), {:symbolize_names => true}))
		end
	
		initdata = initdata.merge(date_key.to_sym => Time.new.strftime('%Y-%m-%d'))
		return initdata
	end

	def deserialize_str(datastr, fmt='yaml', date_key='date')
		initdata = {}
	
		keys2sym = proc {|h| h.map{|k, v|
		  [k.to_sym, v.is_a?(Hash) ? keys2sym.call(v) : v]}.to_h}
	
		if ['yaml', 'json'].include? fmt
			initdata = initdata.merge(Psych.load(datastr, {:symbolize_names => true}))
		elsif 'toml' == fmt
			initdata = initdata.merge(keys2sym.call TOML.load(datastr))
		#elsif 'json' == fmt
		#	#initdata = JSON.parse(datastr, {:symbolize_names => true})
		#	initdata = initdata.merge(JSON.parse(datastr, {:symbolize_names => true}))
		end
	
		initdata = initdata.merge(date_key.to_sym => Time.new.strftime('%Y-%m-%d'))
		return initdata
	end

	def regex_checks(pat, substr, txt)
		if (!(%r"#{pat}".match(substr)))
			puts "ERROR: Regex match failure (%r'#{pat}'.match(#{substr})) for (#{txt})."
			exit 1
		end
	end

	def derive_skel_vars(ctx={})
		name = "#{ctx[:parent]}#{ctx.fetch(:separator, '')}#{ctx[:project]}"
		parentcap = ctx.fetch(:parent, '').split(ctx.fetch(:separator, '-')).map{|s| s.capitalize
			}.join(ctx.fetch(:joiner, ''))
		namespace = "#{ctx[:groupid] ? ctx[:groupid] + '.' : ''}#{ctx.fetch(:parent, '')}.#{ctx[:project]}"
	
		ctx = ctx.merge(:year => ctx[:date].split('-')[0], :name => name,
			:parentcap => parentcap, :projectcap => ctx[:project].capitalize,
			:namespace => namespace, :nesteddirs => namespace.tr('.', '/'))
		return ctx
	end

	def render_skeleton(skeleton='skeleton-ml', ctx={})
		ctx = ctx.merge(:skeletondir => File.expand_path(skeleton, SCRIPTPARENT))
		start_dir = File.join(ctx[:skeletondir], '{{name}}')
		files_skel = Dir["{**/*,**/.*}", base: start_dir
			].filter{|p| File.file? File.join(start_dir, p)}
		renderouts, copyouts, pat_mustache = {}, {}, /\.mustache$/
		files_skel.each{|skelX|
			if pat_mustache.match(skelX)
				renderouts[skelX] = Mustache.render(skelX, ctx).sub(pat_mustache, '')
			else
				copyouts[skelX] = Mustache.render(skelX, ctx)
			end
			}
		puts "... processing files processing -- rendering #{renderouts.size} ; copying #{copyouts.size} ..."
	
		(renderouts.values() + copyouts.values()).uniq.each{|pathX|
			dirX = File.dirname(pathX)
			if not File.exist?(File.join(ctx[:name], dirX))
				FileUtils.mkdir_p(File.join(ctx[:name], dirX))
			end}
		renderouts.each{|srcR, dstR| File.open(File.join(ctx[:name], dstR), 'w+') {|f|
			f.write(Mustache.render(File.read(File.join(start_dir, srcR)), ctx))}}
		copyouts.each{|srcC, dstC| File.open(File.join(ctx[:name], dstC), 'w+') {|f|
			f.write(File.read(File.join(start_dir, srcC)))}}
	
		puts 'Post rendering message'
		Dir.chdir(ctx[:name])
		system('ruby choices/post_gen_project.rb') # ruby ___.rb | sh ___.sh
	end

	def parse_cmdopts(args=[])
		require 'optparse'
	
		opts_hash = {'data': 'data.yaml', 'datafmt': 'yaml', 'fileOut': STDOUT,
		  'template': 'skeleton-ml'}
		usage_str = <<EOF
Usage: #{File.basename $PROGRAM_NAME} [OPTIONS] [TEMPLATE] [KEY1=VAL1 KEYN=VALN]

Example: #{File.basename $PROGRAM_NAME} -d data.yaml LICENSE.mustache
EOF
		opts_parser = OptionParser.new {|opts|
			opts.separator nil
			opts.separator 'Specific options:'
		
			opts.on('-d DATA', '--data DATA', String, 'Data path or - (for stdin)') {
				|data| opts_hash[:data] = data}
			opts.on('-f FMT', '--datafmt FMT', ['yaml', 'json', 'toml'],
				'Specify data file format (yaml, json, toml)') {
				|datafmt| opts_hash[:datafmt] = datafmt}
			#opts.on('-t TEMPLATE', '--template TEMPLATE', String, 'Template path') {
			#	|template| opts_hash[:template] = template}
			#opts.on('-i IN', '--fileIn IN', String, 'File in - (for stdin) or path') {
			#	|fileIn| opts_hash[:fileIn] = File.open(fileIn)}
			opts.on('-o FILEOUT', '--fileOut FILEOUT', String,
			  'File out - (for stdout) or path') {
				|fileOut| opts_hash[:fileOut] = File.open(fileOut, mode = 'w+')}
		
			opts.banner = usage_str
			opts.separator nil
			opts.separator 'Common options:'
		
			opts.on_tail('-h', 'help message') {
				$stderr.print opts
				exit 0 }
		}.parse!(args) or raise
		#raise usage_str unless 0 == args.size
		opts_hash
	end

	def main(args=[])
		opts_hash, cfg = parse_cmdopts(args), {}
		
		kvset, rest = {}, []
		while extra = args.shift
			if extra.match(/\w+=\w+/)
				key1, val1 = *extra.split('=', 2)
				(kvset ||= {})[key1.to_sym] = val1
			else
			  rest += [extra]
			end
		end
		if 0 < rest.length
		  opts_hash[:template] = rest.shift
		end
		if (not File.exist?(opts_hash[:template]) and 
		    not File.exist?(File.expand_path(opts_hash[:template], SCRIPTPARENT)))
		  puts "Non-existent template: #{opts_hash[:template]}"
		  exit 1
		end
		is_dir = (File.directory?(opts_hash[:template]) or 
		  File.directory?(File.expand_path(opts_hash[:template], SCRIPTPARENT)))
	  if '-' == opts_hash[:data]
	    cfg = deserialize_str($stdin.read, opts_hash[:datafmt], 'date')
	  end
	  
	  if not is_dir
	    if not '-' == opts_hash[:data]
	      cfg = deserialize_file(opts_hash[:data], opts_hash[:datafmt], 'date')
	    end
	    cfg = cfg.merge(kvset)
			opts_hash[:fileOut].write(Mustache.render(File.read(opts_hash[:template]), cfg))
	  else
	    if not '-' == opts_hash[:data]
	      cfg = deserialize_file(File.expand_path(
				  "#{opts_hash[:template]}/#{opts_hash[:data]}", SCRIPTPARENT),
				  opts_hash[:datafmt], 'date')
	    end
			cfg = cfg.merge(kvset)
			cfg = derive_skel_vars(cfg)
			regex_checks(cfg.fetch(:parentregex, ''), cfg.fetch(:parent, ''),
				cfg.fetch(:name, ''))
			regex_checks(cfg.fetch(:projectregex, ''), cfg.fetch(:project, ''), 
			  cfg.fetch(:name, ''))
			render_skeleton(opts_hash[:template], cfg)
	  end
		0
	end
end


if __FILE__ == $PROGRAM_NAME
  exit RenderMustache.main(ARGV)
end
